### ISSUE TITLE
pdm add eth-brownie~=1.17 hangs indefinitely

### ISSUE THREAD

**Issue by KholdStare** [pdm-project/pdm#744]
- [x] I have searched the issue tracker and believe that this is not a duplicate.

**Make sure you run commands with `-v` flag before pasting the output.**

## Steps to reproduce

Ran using pdm version 1.10.2

The pyproject.toml:
```
[project]
dependencies = []
requires-python = ">=3.9"

[build-system]
requires = ["pdm-pep517"]
build-backend = "pdm.pep517.api"
```

On commandline:
```
pdm add eth-brownie~=1.17
```

## Actual behavior

The process hangs indefinitely on:

```
Adding packages to default dependencies: eth-brownie
⠋ Resolving: new pin python >=3.9
```

If I use `-v` it goes on forever trying many rounds and backtracking. Here is an excerpt:

```
          <Candidate semantic-version 2.8.5 from https://pypi.org/simple/semantic-version/>
  Adding requirement six==1.16.0(from eth-brownie 1.17.0)
        Found matching candidates:
          <Candidate six 1.16.0 from https://pypi.org/simple/six/>
  Adding requirement sortedcontainers==2.4.0(from eth-brownie 1.17.0)
        Found matching candidates:
          <Candidate sortedcontainers 2.4.0 from https://pypi.org/simple/sortedcontainers/>
  Adding requirement toml==0.10.2(from eth-brownie 1.17.0)
        Found matching candidates:
          <Candidate toml 0.10.2 from https://pypi.org/simple/toml/>
  Adding requirement tomli==1.2.1(from eth-brownie 1.17.0)
        Found matching candidates:
          <Candidate tomli 1.2.1 from https://pypi.org/simple/tomli/>
  Adding requirement toolz==0.11.1(from eth-brownie 1.17.0)
        Found matching candidates:
          <Candidate toolz 0.11.1 from https://pypi.org/simple/toolz/>
  Adding requirement tqdm==4.62.3(from eth-brownie 1.17.0)
        Found matching candidates:
          <Candidate tqdm 4.62.3 from https://pypi.org/simple/tqdm/>
  Adding requirement typing-extensions==3.10.0.2(from eth-brownie 1.17.0)
        Found matching candidates:
          <Candidate typing-extensions 3.10.0.2 from https://pypi.org/simple/typing-extensions/>
  Adding requirement urllib3==1.26.7(from eth-brownie 1.17.0)
        Found matching candidates:
          <Candidate urllib3 1.26.7 from https://pypi.org/simple/urllib3/>
  Adding requirement varint==1.0.2(from eth-brownie 1.17.0)
        Found matching candidates:
          <Candidate varint 1.0.2 from https://pypi.org/simple/varint/>
  Adding requirement vvm==0.1.0(from eth-brownie 1.17.0)
        Found matching candidates:
          <Candidate vvm 0.1.0 from https://pypi.org/simple/vvm/>
  Adding requirement vyper==0.3.0(from eth-brownie 1.17.0)
        Could not find any matching candidates.
        Could not find any matching candidates even when considering pre-releases.
        Found but non-matching candidates:
          <Candidate vyper 0.3.0 from https://pypi.org/simple/vyper/>
          <Candidate vyper 0.2.16 from https://pypi.org/simple/vyper/>
          <Candidate vyper 0.2.15 from https://pypi.org/simple/vyper/>
          <Candidate vyper 0.2.14 from https://pypi.org/simple/vyper/>
          <Candidate vyper 0.2.13 from https://pypi.org/simple/vyper/>
          <Candidate vyper 0.2.12 from https://pypi.org/simple/vyper/>
          <Candidate vyper 0.2.11 from https://pypi.org/simple/vyper/>
          <Candidate vyper 0.2.10 from https://pypi.org/simple/vyper/>
          <Candidate vyper 0.2.9 from https://pypi.org/simple/vyper/>
          <Candidate vyper 0.2.8 from https://pypi.org/simple/vyper/>
          ... [57 more candidate(s)]
Candidate rejected: python None
Backtracking...
======== Ending round 55 ========
```

## Expected behavior

Either succeed, or report an error

## Environment Information

```bash
# Paste the output of `pdm info && pdm info --env` below:
PDM version:        1.10.2
Python Interpreter: /usr/bin/python3.9 (3.9)
Project Root:       /home/kholdstare/workspace/crypto/evm/temp
Project Packages:   /home/kholdstare/workspace/crypto/evm/temp/__pypackages__/3.9
{
  "implementation_name": "cpython",
  "implementation_version": "3.9.2",
  "os_name": "posix",
  "platform_machine": "x86_64",
  "platform_release": "5.10.60.1-microsoft-standard-WSL2",
  "platform_system": "Linux",
  "platform_version": "#1 SMP Wed Aug 25 23:20:18 UTC 2021",
  "python_full_version": "3.9.2",
  "platform_python_implementation": "CPython",
  "python_version": "3.9",
  "sys_platform": "linux"
}
```

**Comment by pawamoy** [comment]
Probably same issue as #627

**Comment by frostming** [comment]
> Probably same issue as #627

I am afraid not, I have submitted a fix in #746, but the bug was introduced by the change of 1.9.0, while #627 is reported on 1.8.0
